### PR TITLE
fix(repeater): resolve bridges service communication 

### LINKED ISSUE #170: Resolve Bridges service communication issues
This issue is to address some `Socket.IO` protocol related discrepancies which are preventing `sectester-net` from successful communication to `Bridges` service.

The scope:

1. Bump `Socket.IO` client to most recent version
2. Eliminate message pack key naming discrepancy
3. Perform protocol message reconciliation using message pack formatters
3.1 to eliminate http headers structure mapping discrepancy
3.2 to allow `HttpMethod` enumeration class serialization 
3.3 to satisfy server requirements for string based native enumeration serialization

### PR DESCRIPTION
**Pull request by ostridm**
closes #170

### CODE REVIEW

File: src/SecTester.Repeater/Bus/HttpMessage.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+{
+  public const string HeadersKey = "headers";
+  public const string BodyKey = "body";
+  public const string ProtocolKey = "protocol";
+
+  [Key(ProtocolKey)]
+  public Protocol Protocol { get; set; } = Protocol.Http;
+
+  [Key(HeadersKey)]
+  [MessagePackFormatter(typeof(MessagePackHttpHeadersFormatter))]
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/HttpMessage.cs]: Why not implement the `IFormatterResolver` interface instead of using the `MessagePackFormatter`?

File: src/SecTester.Repeater/Bus/HttpMessage.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+using System.Collections.Generic;
+using MessagePack;
+using SecTester.Repeater.Bus.Formatters;
+
+namespace SecTester.Repeater.Bus;
+
+public abstract record HttpMessage
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/HttpMessage.cs]: Why do we need a base class for incoming and outgoing? Just to push a few fields up? Please revert these changes as they unnecessarily complicate code without any benefits.

File: src/SecTester.Repeater/Bus/SocketIoMessage.cs [hunk 3]
```diff
@@ -14,7 +16,15 @@ public SocketIoMessage(SocketIOResponse response)
     _response = response ?? throw new ArgumentNullException(nameof(response));
   }
 
-  public virtual T GetValue<T>(int index = 0) => _response.GetValue<T>(index);
+  public virtual T GetValue<T>(int index = 0)
+  {
+    if (typeof(T) == typeof(IncomingRequest))
+    {
+      return (T)(object)IncomingRequest.FromDictionary(_response.GetValue<Dictionary<object, object>>(index));
+    }
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/SocketIoMessage.cs]: It is unclear why a generic `SocketIoMessage` includes such specific details. Why not initially call the `GetValue` with `Dictionary<object, object>` and then map result to `IncomingRequest` in the `request` handler?

File: src/SecTester.Repeater/Bus/Formatters/MessagePackStringEnumFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+using System.Runtime.Serialization;
+using MessagePack;
+using MessagePack.Formatters;
+
+namespace SecTester.Repeater.Bus.Formatters;
+
+// ADHOC: MessagePack-CSharp prohibits declaration of IMessagePackFormatter<T> requesting to use System.Enum instead, refer to formatter interface argument type check
+// https://github.com/MessagePack-CSharp/MessagePack-CSharp/blob/db2320b3338735c9266110bbbfffe63f17dfdf46/src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Resolvers/DynamicObjectResolver.cs#L623
+
+public class MessagePackStringEnumFormatter<T> : IMessagePackFormatter<Enum>
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackStringEnumFormatter.cs]:
```suggestion
public class MessagePackStringEnumMemberFormatter<T> : IMessagePackFormatter<Enum>
```

File: src/SecTester.Repeater/Protocol.cs [hunk 5]
```diff
@@ -1,6 +1,9 @@
+using System.Runtime.Serialization;
+
 namespace SecTester.Repeater;
 
 public enum Protocol
 {
+  [EnumMember(Value = "http")]
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Protocol.cs]: Allow passing a naming strategy to the formatter to avoid redundancy
```c#
[MessagePackFormatter(typeof(MessagePackStringEnumFormatter), new MessagePackStringEnumFormatterOptions { NamingPolicy = SnakeCaseNamingPolicy.Instance })]
```

File: src/SecTester.Repeater/Bus/Formatters/MessagePackStringEnumFormatter.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    writer.Write(stringValue);
+  }
+
+  public Enum Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+  {
+
+    var stringValue = reader.ReadString();
+
+    if (!StringToEnum.TryGetValue(stringValue, out var enumValue))
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackStringEnumFormatter.cs]: Possible null reference  (for details please see the workflow annotations)

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpMethodFormatter.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+  public HttpMethod? Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+  {
+    switch (reader.NextMessagePackType)
+    {
+      case MessagePackType.Nil:
+        return null;
+      case MessagePackType.String:
+        var method = reader.ReadString();
+        return string.IsNullOrWhiteSpace(method) ? null : new HttpMethod(method);
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpMethodFormatter.cs]: Do not create new instance of `HttpMethod`; treat it like a registry. Refer to approach used in `JsonHttpMethodEnumerationStringConverter`

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    finally
+    {
+      reader.Depth--;
+    }
+
+    return result;
+  }
+
+  private static string DeserializeString(ref MessagePackReader reader)
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: If header field value is null, we should simply fallback to empty string instead of raising exception.

File: src/SecTester.Repeater/Bus/IncomingRequest.cs [hunk 9]
```diff
@@ -1,19 +1,93 @@
...
+  [MessagePackFormatter(typeof(MessagePackHttpMethodFormatter))]
   public HttpMethod Method { get; set; } = HttpMethod.Get;
-  public Protocol Protocol { get; set; } = Protocol.Http;
+
+  [Key(UrlKey)]
   public Uri Url { get; set; } = Url ?? throw new ArgumentNullException(nameof(Url));
-  public IEnumerable<KeyValuePair<string, IEnumerable<string>>> Headers { get; set; } =
-    new List<KeyValuePair<string, IEnumerable<string>>>();
+
+  public static IncomingRequest FromDictionary(Dictionary<object, object> dictionary)
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/IncomingRequest.cs]: Why do not map fields automatically like this:

```csharp
[Key(HeadersKey)]
[MessagePackFormatter(typeof(MessagePackHttpHeadersFormatter))]
private IEnumerable<KeyValuePair<string, object?>> _headers;

public IEnumerable<KeyValuePair<string, IEnumerable<string>>> Headers
{
    get => _headers.Select(kvp => kvp.Value switch
    {
        IEnumerable<string> strings => new KeyValuePair<string, IEnumerable<string>>(kvp.Key, strings),
        null => new KeyValuePair<string, IEnumerable<string>>(kvp.Key, Enumerable.Empty<string>()),
        _ => new KeyValuePair<string, IEnumerable<string>>(kvp.Key, new[] { kvp.Value.ToString() })
    });
    set => _headers = value.Select(kvp => new KeyValuePair<string, object?>(kvp.Key, kvp.Value?.ToArray()));
}
``` 

The same could be applied to the rest of the fields.

File: src/SecTester.Repeater/Bus/Formatters/MessagePackStringEnumFormatter.cs
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackStringEnumFormatter.cs:1]: Please move all formatters to the Internal folder; they are not intended for public use.

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+using MessagePack;
+using MessagePack.Formatters;
+
+namespace SecTester.Repeater.Bus.Formatters;
+
+// Headers formatter is to be support javascript `undefined` which is treated as null (0xC0)
+// https://www.npmjs.com/package/@msgpack/msgpack#messagepack-mapping-table
+// https://github.com/msgpack/msgpack/blob/master/spec.md#nil-format
+
+public class MessagePackHttpHeadersFormatter : IMessagePackFormatter<
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: You can simply register formatter for `KeyValuePair<string, IEnumerable<string>>` to avoid redundant complexity and duplicating the built-in logic.

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    {
+      return DeserializeMap(ref reader, length, options);
+    }
+    finally
+    {
+      reader.Depth--;
+    }
+  }
+
+  private static void SerializeMap(ref MessagePackWriter writer, IEnumerable<KeyValuePair<string, IEnumerable<string>>> value)
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: Let's use method overloading for clarity whenever possible.

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  private static IEnumerable<string> DeserializeArray(ref MessagePackReader reader, int length, MessagePackSerializerOptions options)
+  {
+    var result = new List<string>(length);
+
+    if (length == 0)
+    {
+      return result;
+    }
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: Avoid redundant conditions and branches; they only complicate the code w/o reason

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+      var key = DeserializeString(ref reader);
+
+      switch (reader.NextMessagePackType)
+      {
+        case MessagePackType.String:
+          result.Add(new KeyValuePair<string, IEnumerable<string>>(key, new List<string> { DeserializeString(ref reader) }));
+          break;
+        case MessagePackType.Array:
+          result.Add(new KeyValuePair<string, IEnumerable<string>>(key, DeserializeArray(ref reader, reader.ReadArrayHeader(), options)));
+          break;
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: Extract this into `DeserializeValue` method, similar to how you did with `SerializeValue`

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    }
+  }
+
+  public IEnumerable<KeyValuePair<string, IEnumerable<string>>>? Deserialize(ref MessagePackReader reader,
+    MessagePackSerializerOptions options)
+  {
+    switch (reader.NextMessagePackType)
+    {
+      case MessagePackType.Nil:
+        return null;
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: You should read anyway to advance the reader

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+      case MessagePackType.Map:
+        break;
+      default:
+        throw new MessagePackSerializationException(string.Format(CultureInfo.InvariantCulture,
+          "Unrecognized code: 0x{0:X2} but expected to be a map or null", reader.NextCode));
+    }
```
Comment (derevnjuk) [inline src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]:
```suggestion
           if (reader.NextMessagePackType == MessagePackType.Nil) {
                reader.ReadNil();
                return null;
            }
            if (reader.NextMessagePackType != MessagePackType.Map) {
                throw new MessagePackSerializationException(string.Format(CultureInfo.InvariantCulture,
                    "Unrecognized code: 0x{0:X2} but expected to be a map or null", reader.NextCode));
            }
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: src/SecTester.Repeater/Bus/IncomingRequest.cs (hunk 9 above)
Comment (ostridm) [reply src/SecTester.Repeater/Bus/IncomingRequest.cs]: Looks promising, but this is not going to work. The SocketIO `PackMessage` is not compatible to `*AllowPrivate` resolvers.

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs (hunk 10 above)
Comment (ostridm) [reply src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: It's need to convert map to array and vice versa.

**Review by derevnjuk — DISMISSED** [review]

File: src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs (hunk 10 above)
Comment (derevnjuk) [reply src/SecTester.Repeater/Bus/Formatters/MessagePackHttpHeadersFormatter.cs]: Not clear what you mean. Anyway,`Dictionary` also implements `IEnumerable`, just like `List`. `IEnumerable<KeyValuePair<string, IEnumerable<string>>>` is equivalent to `IDictionary<string, IEnumerable<string>>`
